zerver/actions/user_groups.py

```diff
@@ -271,51 +271,75 @@ def do_send_subgroups_update_event(
 
 @transaction.atomic
 def add_subgroups_to_user_group(
-    user_group: UserGroup, subgroups: List[UserGroup], *, acting_user: Optional[UserProfile]
+    supergroup: UserGroup, subgroups: List[UserGroup], *, acting_user: Optional[UserProfile]
 ) -> None:
     group_memberships = [
-        GroupGroupMembership(supergroup=user_group, subgroup=subgroup) for subgroup in subgroups
+        GroupGroupMembership(supergroup=supergroup, subgroup=subgroup) for subgroup in subgroups
     ]
     GroupGroupMembership.objects.bulk_create(group_memberships)
 
     subgroup_ids = [subgroup.id for subgroup in subgroups]
     now = timezone_now()
     RealmAuditLog.objects.bulk_create(
         RealmAuditLog(
-            realm=user_group.realm,
-            modified_user_group=user_group,
-            event_type=RealmAuditLog.USER_GROUP_DIRECT_SUBGROUP_MEMBERSHIP_ADDED,
+            realm=supergroup.realm,
+            modified_user_group_id=modified_uer_group_id,
+            event_type=event_type,
             event_time=now,
             acting_user=acting_user,
-            extra_data=orjson.dumps({"subgroup_id": subgroup_id}).decode(),
+            extra_data=extra_data,
         )
         for subgroup_id in subgroup_ids
+        for event_type, extra_data, modified_uer_group_id in (
+            (
+                RealmAuditLog.USER_GROUP_DIRECT_SUBGROUP_MEMBERSHIP_ADDED,
+                orjson.dumps({"subgroup_id": subgroup_id}).decode(),
+                supergroup.id,
+            ),
+            (
+                RealmAuditLog.USER_GROUP_DIRECT_SUPERGROUP_MEMBERSHIP_ADDED,
+                orjson.dumps({"supergroup_id": supergroup.id}).decode(),
+                subgroup_id,
+            ),
+        )
```

Comment:
I am finding this difficult to read and similar for changes in `remove_subgroups_from_user_group`. Would probably be better to have something like what you have done in `create_system_user_groups_for_realm` -
```python
realmauditlog_objects = []
for subgroup in subgroup_ids:
    realmauditlog_objects.extend(
        [
            RealmAuditLog(
                realm=supergroup.realm,
                modified_user_group_id=supergroup.id,
                event_type=RealmAuditLog.USER_GROUP_DIRECT_SUBGROUP_MEMBERSHIP_REMOVED,
                event_time=now,
                acting_user=acting_user,
                extra_data=orjson.dumps({"subgroup_id": subgroup_id}).decode(),
            ),
             RealmAuditLog(
                realm=supergroup.realm,
                modified_user_group_id=subgroup_id,
                event_type=RealmAuditLog.USER_GROUP_DIRECT_SUPERGROUP_MEMBERSHIP_REMOVED,
                event_time=now,
                acting_user=acting_user,
                extra_data=orjson.dumps({"supergroup_id": supergroup.id}).decode(),
            )
        ]
    )
RealmAuditLog.objects.bulk_create(...)
```